File: core/src/galenscovell/oregontrail/util/Repository.java
package galenscovell.oregontrail.util;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import galenscovell.oregontrail.map.Tile;
import galenscovell.oregontrail.things.inanimate.Location;
import galenscovell.oregontrail.ui.components.GameStage;
import galenscovell.oregontrail.ui.screens.GameScreen;

import java.util.*;

public class Repository {
    public static GameScreen gameScreen;
    public static ArrayList<Location> locations;
    public static List<Location> locationsInRange;
    public static Location currentLocation;
    public static Location currentSelection;
    public static Tile[][] grid;
    public static ShapeRenderer shapeRenderer;
    public static int playerRange;

    private Repository() {}


    /**************************
     * Called from GameScreen
     */
    public static void setup(GameScreen game) {
        gameScreen = game;
        shapeRenderer = new ShapeRenderer();
        playerRange = 12;
    }

    public static void setTargetsInRange() {
        locationsInRange = new ArrayList<Location>();

        for (Location location : locations) {
            double squareDist = Math.pow(currentLocation.getTile().x - location.getTile().x, 2) + Math.pow(currentLocation.getTile().y - location.getTile().y, 2);

            if (squareDist <= Math.pow(playerRange, 2)) {
                locationsInRange.add(location);
            }
        }
    }

    public static void drawShapes() {
        float radius = playerRange * Constants.TILESIZE;
        float centerX = currentLocation.getTile().x * Constants.TILESIZE + (Constants.TILESIZE / 2);
        float centerY = Gdx.graphics.getHeight() - (currentLocation.getTile().y * Constants.TILESIZE) - (2 * Constants.TILESIZE) - (Constants.TILESIZE / 2);

        shapeRenderer.begin(ShapeRenderer.ShapeType.Line);
        // Circle rendering
        shapeRenderer.setColor(0.95f, 0.61f, 0.07f, 0.6f);
        shapeRenderer.circle(centerX, centerY, radius);
        shapeRenderer.circle(centerX, centerY, 20);

        // Path rendering
        if (locationsInRange != null && locationsInRange.size() > 0) {
            shapeRenderer.setColor(0.93f, 0.94f, 0.95f, 0.6f);
            for (Location location : locationsInRange) {
                shapeRenderer.line(
                            (currentLocation.getTile().x * Constants.TILESIZE) + (Constants.TILESIZE / 2),
                            Gdx.graphics.getHeight() - (currentLocation.getTile().y * Constants.TILESIZE) - (2 * Constants.TILESIZE) - (Constants.TILESIZE / 2),
                            (location.getTile().x * Constants.TILESIZE) + (Constants.TILESIZE / 2),
                            Gdx.graphics.getHeight() - (location.getTile().y * Constants.TILESIZE) - (2 * Constants.TILESIZE) - (Constants.TILESIZE / 2)
                    );
            }
        }
        // Selection rendering
        if (currentSelection != null) {
            shapeRenderer.setColor(0.18f, 0.8f, 0.44f, 0.6f);
            float selectionX = currentSelection.getTile().x * Constants.TILESIZE + (Constants.TILESIZE / 2);
            float selectionY = Gdx.graphics.getHeight() - (currentSelection.getTile().y * Constants.TILESIZE) - (2 * Constants.TILESIZE) - (Constants.TILESIZE / 2);
            shapeRenderer.circle(selectionX, selectionY, 20);
        }
        shapeRenderer.end();
    }


    /**************************
     * Called from MapPanel
     */
    public static void setTiles(Tile[][] tiles) {
        grid = tiles;
    }

    public static boolean travelToSelection() {
        if (currentSelection != null && locationsInRange.contains(currentSelection)) {
            currentLocation.getTile().becomeExplored();
            currentLocation = currentSelection;
            currentSelection.getTile().becomeCurrent();
            currentSelection.enter();
            setSelection(null);
            return true;
        }
        return false;
    }


    /**************************
     * Called from Tile
     */
    public static void setSelection(Tile selection) {
        if (selection == null) {
            GameStage gameStage = (GameStage) gameScreen.getGameStage();
            gameStage.updateDistanceLabel("Distance: 0.0 AU");
            return;
        }

        for (Location location : locations) {
            if (location.getTile() == selection) {
                if (!(location == currentLocation)) {
                    currentSelection = location;

                    double distance = Math.sqrt(Math.pow(currentLocation.getTile().x - selection.x, 2) + Math.pow(currentLocation.getTile().y - selection.y, 2)) * 4;
                    GameStage gameStage = (GameStage) gameScreen.getGameStage();
                    gameStage.updateDistanceLabel("Distance: " + String.format("%.1f", distance) + " AU");
                }
            }
        }
    }


    /**************************
     * Called from MapGenerator
     */
    public static void populateLocations(ArrayList<Location> locationsToSet) {
        locations = locationsToSet;
        currentLocation = null;
        for (Location location : locations) {
            if (currentLocation == null || location.x < currentLocation.x) {
                currentLocation = location;
            }
        }
        currentLocation.getTile().becomeCurrent();
    }
}

File: core/src/galenscovell/hinterstar/util/Repository.scala
package galenscovell.hinterstar.util

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import galenscovell.hinterstar.map.Sector
import galenscovell.hinterstar.processing.{EventContainer, EventParser}
import galenscovell.hinterstar.things.inanimate.Location
import galenscovell.hinterstar.ui.components.GameStage
import galenscovell.hinterstar.ui.screens.GameScreen

import scala.collection.mutable.ArrayBuffer


object Repository {
  var gameScreen: GameScreen = null
  var locations: ArrayBuffer[Location] = ArrayBuffer()
  var locationsInRange: ArrayBuffer[Location] = ArrayBuffer()
  var currentLocation: Location = null
  var currentSelection: Location = null
  var sectors: Array[Array[Sector]] = null
  var shapeRenderer: ShapeRenderer = null
  var playerRange: Int = 0
  val eventParser: EventParser = new EventParser()


  /**
    * Called from GameScreen
    */
  def setup(game: GameScreen): Unit = {
    gameScreen = game
    shapeRenderer = new ShapeRenderer
    playerRange = 10
  }

  def setTargetsInRange(): Unit = {
    locationsInRange.clear()

    for (location <- locations) {
      val squareDist: Double = Math.pow(currentLocation.getSector.sx - location.getSector.sx, 2) + Math.pow(currentLocation.getSector.sy - location.getSector.sy, 2)
      if (squareDist <= Math.pow(playerRange, 2)) {
        locationsInRange += location
      }
    }
  }

  def drawShapes(): Unit = {
    val radius: Float = playerRange * Constants.SECTORSIZE
    val centerX: Float = currentLocation.getSector.sx * Constants.SECTORSIZE + (Constants.SECTORSIZE / 2)
    val centerY: Float = Gdx.graphics.getHeight - (currentLocation.getSector.sy * Constants.SECTORSIZE) - (2 * Constants.SECTORSIZE) - (Constants.SECTORSIZE / 2)

    shapeRenderer.begin(ShapeRenderer.ShapeType.Line)
    shapeRenderer.setColor(0.95f, 0.61f, 0.07f, 0.6f)
    shapeRenderer.circle(centerX, centerY, radius)
    shapeRenderer.circle(centerX, centerY, 20)

    if (locationsInRange != null && locationsInRange.size > 0) {
      shapeRenderer.setColor(0.93f, 0.94f, 0.95f, 0.6f)
      for (location <- locationsInRange) {
        shapeRenderer.line(
          (currentLocation.getSector.sx * Constants.SECTORSIZE) + (Constants.SECTORSIZE / 2),
          Gdx.graphics.getHeight - (currentLocation.getSector.sy * Constants.SECTORSIZE) - (2 * Constants.SECTORSIZE) - (Constants.SECTORSIZE / 2),
          (location.getSector.sx * Constants.SECTORSIZE) + (Constants.SECTORSIZE / 2),
          Gdx.graphics.getHeight - (location.getSector.sy * Constants.SECTORSIZE) - (2 * Constants.SECTORSIZE) - (Constants.SECTORSIZE / 2)
        )
      }
    }
    if (currentSelection != null) {
      shapeRenderer.setColor(0.18f, 0.8f, 0.44f, 0.6f)
      val selectionX: Float = currentSelection.getSector.sx * Constants.SECTORSIZE + (Constants.SECTORSIZE / 2)
      val selectionY: Float = Gdx.graphics.getHeight - (currentSelection.getSector.sy * Constants.SECTORSIZE) - (2 * Constants.SECTORSIZE) - (Constants.SECTORSIZE / 2)
      shapeRenderer.circle(selectionX, selectionY, 20)
    }
    shapeRenderer.end()
  }


  /**
    * Called from GameStage
    */
  def parseNextEvent: EventContainer = {
    val nextEventType: String = currentLocation.getCurrentEvent.getType
    val parsedEvent: EventContainer = eventParser.parse(nextEventType)
    parsedEvent
  }


  /**
    * Called from MapPanel
    */
  def setSectors(sectors: Array[Array[Sector]]): Unit = {
    Repository.sectors = sectors
  }

  def travelToSelection: Boolean = {
    if (currentSelection != null && locationsInRange.contains(currentSelection)) {
      currentLocation.getSector.becomeExplored()
      currentLocation = currentSelection
      currentSelection.getSector.becomeCurrent()
      currentSelection.enter()
      setSelection(null)
      true
    } else {
      false
    }
  }


  /**
    * Called from Sector
    */
  def setSelection(selection: Sector): Unit = {
    if (selection == null) {
      val gameStage: GameStage = gameScreen.getGameStage.asInstanceOf[GameStage]
      gameStage.updateDistanceLabel("Distance: 0.0 AU")
    } else {
      for (location <- locations) {
        if (location.getSector eq selection) {
          if (!(location eq currentLocation)) {
            currentSelection = location
            val distance: Double = Math.sqrt(Math.pow(currentLocation.getSector.sx - selection.sx, 2) + Math.pow(currentLocation.getSector.sy - selection.sy, 2)) * 4
            val gameStage: GameStage = gameScreen.getGameStage.asInstanceOf[GameStage]
            gameStage.updateDistanceLabel(f"Distance: $distance%1.1f AU")
          }
        }
      }
    }
  }


  /**
    * Called from MapGenerator
    */
  def populateLocations(locationsToSet: ArrayBuffer[Location]): Unit = {
    locations = locationsToSet
    for (location <- locations) {
      if (currentLocation == null || location.x < currentLocation.x) {
        currentLocation = location
      }
    }
    currentLocation.getSector.becomeCurrent()
  }
}
